marshal/admin.go

```diff
@@ -155,35 +155,57 @@ func (a *consumerGroupAdmin) releaseClaims(resetOffset bool) error {
 	}
 }
 
-// heartbeatLoop hearbeats as if we had a claim to this partition and were simply
+// Send a single heartbeat using the given offset.
+func (a *consumerGroupAdmin) heartbeat(
+	topic string, partID int, offset int64) (ok bool) {
+	// If we fail to heartbeat, record this in claimHealth.
+	// The Admin will take care of cleaning up other claims.
+	if err := a.marshaler.Heartbeat(topic, partID, offset); err != nil {
+		log.Errorf("[%s:%d] Admin failed to heartbeat. It is now unhealthy "+
+			"and will not reset offsets.", topic, partID)
+		atomic.StoreInt32(a.claimHealth, 0)
+		return false
+	}
+	return true
+}
+
+// heartbeatLoop heartbeats as if we had a claim to this partition and were simply
 // not reading past where the previous owner had left off.
+//
+// Callers can signal for the heartbeatLoop to stop by sending a message on the
+// stopHeartbeat channel; the heartbeatLoop will close the heartbeatStopped
+// channel after it has in fact stopped.
 func (a *consumerGroupAdmin) heartbeatLoop(
-	topic string, partID int, lastOffset int64, stopChan chan struct{}) {
-
+	topic string, partID int, lastOffset int64,
+	stopHeartbeat, heartbeatStopped chan struct{}) {
```

Comment:
Yeah, more than 100.  BTW I started a new line for the first parameter to adhere to our company golang style guide, but feel free to let me know if you prefer a different convention for Marshal.